### ISSUE TITLE
incorrect plt populated for an ARM binary

### ISSUE THREAD

**Issue by ripples-alive** [Gallopsled/pwntools#614]
I compiled a hello world program with gcc on debian for ARM.
For this binary, the plt populated by pwntools is partially incorrect.

The result for `objdump -d -j .plt a.out`:

```
Disassembly of section .plt:

000082d8 <puts@plt-0x14>:
    82d8:	e52de004 	push	{lr}		; (str lr, [sp, #-4]!)
    82dc:	e59fe004 	ldr	lr, [pc, #4]	; 82e8 <_init+0x1c>
    82e0:	e08fe00e 	add	lr, pc, lr
    82e4:	e5bef008 	ldr	pc, [lr, #8]!
    82e8:	0000823c 	.word	0x0000823c

000082ec <puts@plt>:
    82ec:	e28fc600 	add	ip, pc, #0, 12
    82f0:	e28cca08 	add	ip, ip, #8, 20	; 0x8000
    82f4:	e5bcf23c 	ldr	pc, [ip, #572]!	; 0x23c

000082f8 <__libc_start_main@plt>:
    82f8:	e28fc600 	add	ip, pc, #0, 12
    82fc:	e28cca08 	add	ip, ip, #8, 20	; 0x8000
    8300:	e5bcf234 	ldr	pc, [ip, #564]!	; 0x234

00008304 <__gmon_start__@plt>:
    8304:	4778      	bx	pc
    8306:	46c0      	nop			; (mov r8, r8)
    8308:	e28fc600 	add	ip, pc, #0, 12
    830c:	e28cca08 	add	ip, ip, #8, 20	; 0x8000
    8310:	e5bcf228 	ldr	pc, [ip, #552]!	; 0x228

00008314 <abort@plt>:
    8314:	e28fc600 	add	ip, pc, #0, 12
    8318:	e28cca08 	add	ip, ip, #8, 20	; 0x8000
    831c:	e5bcf220 	ldr	pc, [ip, #544]!	; 0x220
```

The plt populated by pwntools:

```
__gmon_start__ : 0x8304
abort : 0x8310
__libc_start_main : 0x82f8
puts : 0x82ec
```

The plt for `abort` should be 0x8314, but recognised as 0x8310.

**Comment by zachriggle** [comment]
Can you upload the binary, just to make sure we're testing against the same
thing?

**Comment by ripples-alive** [comment]
Of course, but I just don't know how to upload the binary.

**Comment by ripples-alive** [comment]
Oh, I found that it's ok to attach files in comments. 
Here is the binary: [hello.zip](https://github.com/Gallopsled/pwntools/files/212251/hello.zip)

**Comment by zachriggle** [comment]
Looks like the issue is that the `__gmon_start__` PLT entry is in Thumb mode.

I'm not sure of a good way to check that outside of looking for the `bx pc` opcode.

**Comment by zachriggle** [comment]
Fixed in https://github.com/binjitsu/binjitsu/commit/b868c12d6347603092abcc3cdb6b6cb5b7a32c76

**Comment by ripples-alive** [comment]
Yeah, the `bx pc` stub is a tail call optimise did by gcc as there's no jump-to-offset-and-switch-mode instruction (bx jumps to an address in a register, not an immediate offset).
When I apply `-O2` for gcc, I can find more plt entries that have this stub.
The fix by looking for `bx pc` seems not graceful, but it does work.

**Comment by Idolf** [comment]
This was fixed by #636.
